frontend/src/main/java/com/google/tripmeout/frontend/storage/PersistentTripStorage.java

```diff
@@ -0,0 +1,95 @@
+package com.google.tripmeout.frontend.storage;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.tripmeout.frontend.TripModel;
+import com.google.tripmeout.frontend.error.TripAlreadyExistsException;
+import com.google.tripmeout.frontend.error.TripNotFoundException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class PersistentTripStorage implements TripStorage {
+  private static final String TRIP_ID_PROPERTY_NAME = "tripId";
+  private static final String USER_ID_PROPERTY_NAME = "userId";
+  private static final String NAME_PROPERTY_NAME = "name";
+  private static final String PLACE_API_ID_PROPERTY_NAME = "placeId";
+
+  private final DatastoreService datastore;
+
+  PersistentTripStorage(DatastoreService datastore) {
+    this.datastore = datastore;
+  }
+
+  public void addTrip(TripModel trip) throws TripAlreadyExistsException {
+    Key tripKey = KeyFactory.stringToKey(trip.id());
+    try {
+      datastore.get(tripKey);
+      throw new TripAlreadyExistsException("Trip Already Exits");
+    } catch (EntityNotFoundException e) {
+      Entity tripEntity = new Entity(tripKey);
+      tripEntity.setProperty(TRIP_ID_PROPERTY_NAME, trip.id());
+      tripEntity.setProperty(USER_ID_PROPERTY_NAME, trip.userId());
+      tripEntity.setProperty(NAME_PROPERTY_NAME, trip.name());
+      tripEntity.setProperty(PLACE_API_ID_PROPERTY_NAME, trip.placesApiPlaceId());
+      datastore.put(tripEntity);
+    }
+  }
+  public void removeTrip(String tripId) throws TripNotFoundException {
+    Key tripKey = KeyFactory.stringToKey(tripId);
+    try {
+      datastore.get(tripKey);
```

Comment:
If we're going to bother throwing an error on not-found, these need to be done within a transaction.